src/charm.py

```diff
@@ -0,0 +1,323 @@
+#!/usr/bin/env python3
+# Copyright 2023 Mariyan Dimitrov
+# See LICENSE file for licensing details.
+#
+# Learn more at: https://juju.is/docs/sdk
+
+"""Charm the service.
+
+Refer to the following post for a quick-start guide that will help you
+develop a new k8s charm using the Operator Framework:
+
+https://discourse.charmhub.io/t/4208
+"""
+
+import collections
+import logging
+import os
+
+import ops.lib
+from charms.nginx_ingress_integrator.v0.nginx_route import require_nginx_route
+from ops.charm import CharmBase
+from ops.framework import StoredState
+from ops.main import main
+from ops.model import ActiveStatus, WaitingStatus
+
+# Log messages can be retrieved using juju debug-log
+logger = logging.getLogger(__name__)
+
+pgsql = ops.lib.use("pgsql", 1, "postgresql-charmers@lists.launchpad.net")
+DATABASE_NAME = "synapse"
+
+VALID_LOG_LEVELS = ["info", "debug", "warning", "error", "critical"]
+
+
+class MatrixOperatorCharm(CharmBase):
+    """Charm the service."""
+
+    _CONTAINER_NAME = "synapse"
+    _SYNAPSE_CONFIG_PATH = "/data/homeserver.yaml"
+
+    _stored = StoredState()
+
+    def __init__(self, *args):
+        super().__init__(*args)
+        self.framework.observe(self.on.install, self._on_install)
```

Comment:
While observing on.install, I got an error in line `process = self._container().exec(` because the container was not ready yet. After removing it, the deploy was successful. Is there a reason for observing this event?

Reply:
My intent is to run a generate for the template once and create a `homeserver.yaml` whenever a config changes, my understanding was that I could use `on.install` for a one-time initial setup like for the generate function. There might be a better way, as I see that `on.install` relates to unit setup and not to charm setup, I might've been lucky with the ordering as I haven't faced the error. I think it's safe to remove it and rely on the `on.start` or consider moving the logic to `on.config_changed` altogether.